test/functional/feature_bind_port_externalip.py

```diff
@@ -0,0 +1,76 @@
+# Linux:
+# ifconfig lo:0 1.1.1.1/32 up  # to set up
+# ifconfig lo:0 down  # to remove it, after the test
+# FreeBSD:
+# ifconfig lo0 1.1.1.1/32 alias  # to set up
+# ifconfig lo0 1.1.1.1 -alias  # to remove it, after the test
+addr = '1.1.1.1'
+
+# array of tuples [arguments, expected port in localaddresses]
+expected = [
+    [['-externalip=2.2.2.2',       '-port=20001'                               ], 20001],
+    [['-externalip=2.2.2.2',       '-port=20002', '-bind={}'.format(addr)],       20002],
+    [['-externalip=2.2.2.2',                      '-bind={}'.format(addr)],       11002],
+    [['-externalip=2.2.2.2',       '-port=20003', '-bind={}:20004'.format(addr)], 20004],
+    [['-externalip=2.2.2.2',                      '-bind={}:20005'.format(addr)], 20005],
+    [['-externalip=2.2.2.2:20006', '-port=20007'                               ], 20006],
+    [['-externalip=2.2.2.2:20008', '-port=20009', '-bind={}'.format(addr)],       20008],
+    [['-externalip=2.2.2.2:20010',                '-bind={}'.format(addr)],       20010],
+    [['-externalip=2.2.2.2:20011', '-port=20012', '-bind={}:20013'.format(addr)], 20011],
+    [['-externalip=2.2.2.2:20014',                '-bind={}:20015'.format(addr)], 20014],
+    [['-externalip=2.2.2.2',       '-port=20016', '-bind={}:20017'.format(addr),
+                                             '-whitebind={}:20018'.format(addr)], 20017],
+    [['-externalip=2.2.2.2',       '-port=20019',
+                                             '-whitebind={}:20020'.format(addr)], 20020],
+]
+
+class BindPortTest(BitcoinTestFramework):
+    def set_test_params(self):
+        # Avoid any -bind= on the command line. Force the framework to avoid adding -bind=127.0.0.1.
+        self.setup_clean_chain = True
+        self.bind_to_localhost_only = False
+        self.num_nodes = len(expected)
+
+    def add_options(self, parser):
+        parser.add_argument(
+            "--ihave1111", action='store_true', dest="ihave1111",
+            help="Run the test, assuming {} is configured on the machine".format(addr),
+            default=False)
+
+    def setup_nodes(self):
+        if not self.options.ihave1111:
+            raise SkipTest(
+                "To run this test make sure that {} (a routable address) is assigned "
+                "to one of the interfaces on this machine and rerun with --ihave1111".format(addr))
+        args = []
+        for t in expected:
+            args.append(t[0])
+        self.add_nodes(self.num_nodes, extra_args=args)
+        self.start_nodes()
+
+    def run_test(self):
+        for i in range(len(expected)):
+            expected_port = expected[i][1]
+            found = False
+            for local in self.nodes[i].getnetworkinfo()['localaddresses']:
+                if local['address'] == '2.2.2.2':
+                    assert_equal(local['port'], expected_port)
+                    found = True
+                    break
+            assert(found)
```

Comment:
I believe this can be written shorter

```suggestion
            local = next(l for l in self.nodes[i].getnetworkinfo()['localaddresses'] if l['address'] == '2.2.2.2')
            assert_equal(local['port'], expected_port)
```

Reply:
To me that is less readable. Maybe my python level is poor, but then maybe other readers of this are also poor pythoners. I prefer to keep it simple stupid:
```
for ...
    if ...
        break
```
that is nearly the same in any programming language.